homeassistant/components/notify/xmpp.py

```diff
@@ -93,29 +118,217 @@ def __init__(self):
+                url = await self.upload_file_from_url(
+                    get_url, timeout=timeout)
+            elif data.get(ATTR_URL):
+                url = await self.upload_file_from_url(
+                    data.get(ATTR_URL), timeout=timeout)
+            elif data.get(ATTR_PATH_TEMPLATE):
+                _LOGGER.debug(
+                    "Got path template: %s", data.get(ATTR_PATH_TEMPLATE))
+                templ = template_helper.Template(
+                    data.get(ATTR_PATH_TEMPLATE), hass)
+                get_path = template_helper.render_complex(templ, None)
+                url = await self.upload_file_from_path(
+                    get_path, timeout=timeout)
+            elif data.get(ATTR_PATH):
+                url = await self.upload_file_from_path(
+                    data.get(ATTR_PATH), timeout=timeout)
+            else:
+                url = None
+
+            if url is None:
+                _LOGGER.error("No path or URL found for file")
+
+            return url
+
+        async def upload_file_from_url(self, url, timeout=None):
+            """Upload a file from a URL. Returns a URL.
+
+            uploaded via XEP_0363 and HTTP and returns the resulting URL
+            """
+            _LOGGER.info("Getting file from %s", url)
+
+            def get_url(url):
+                return requests.get(
+                    url, verify=data.get(ATTR_VERIFY, True), timeout=timeout)
+            result = await hass.async_add_executor_job(get_url, url)
+
+            if result.status_code >= 400:
+                _LOGGER.error("Could not load file from %s", url)
+                return None
+
+            filesize = len(result.content)
+
+            # we need a file extension, the upload server needs a
+            # filename, if none is provided, through the path we guess
+            # the extension
+            # also setting random filename for privacy
+            if data.get(ATTR_PATH):
+                # using given path as base for new filename. Don't guess type
+                filename = self.get_random_filename(data.get(ATTR_PATH))
+            else:
+                extension = mimetypes.guess_extension(
+                    result.headers['Content-Type']) or ".unknown"
+                _LOGGER.debug("Got %s extension", extension)
+                filename = self.get_random_filename(extension)
+
+            _LOGGER.info("Uploading file from URL, %s", filename)
+
+            url = await self['xep_0363'].upload_file(
+                filename, size=filesize, input_file=result.content,
+                content_type=result.headers['Content-Type'], timeout=timeout)
+
+            return url
+
+        async def upload_file_from_path(self, path, timeout=None):
+            """Upload a file from a local file path via XEP_0363."""
+            _LOGGER.info('Uploading file from path, %s ...', path)
+
+            with open(path, 'rb') as upfile:
```

Comment:
Maybe we should check for whitelisted paths with `hass.config.is_allowed_path(path)`?

I am not sure when that is supposed to be used.

Reply:
YES! thank you. good catch. We should enforce the whitelist here.